### PR TITLE
Introducing scaling function in copy builders

### LINKED ISSUE #1011: Implement scaling function in system participant inputs
The current method of scaling system participants in simona is not sufficient, as storage capacities etc. are not scaled, while output power is.
A convenient way of scaling participants would be to offer a method `SystemParticipantInput scale(Double factor)` that returns a scaled copy of itself.

### PR DESCRIPTION
**Pull request by sebastian-peter**
Closes #1011

Also implementing a bunch of missing copy builders

### CODE REVIEW

**Comment by sebastian-peter** [comment]
@reviewers: I'd be happy if you took a closer look into what parameters I do scale and I don't scale here and whether you agree with me!

**Comment by danielfeismann** [comment]
A more general question. How should we deal with capex? I guess we should scale it for all models as well even this does not represent economy of scale...

File: src/test/groovy/edu/ie3/datamodel/models/input/system/type/chargingpoint/ChargingPointTypeTest.groovy [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+class ChargingPointTypeTest extends Specification {
+
+  def "A ChargingPointType copy method should work as expected"() {
+    given:
+    def cpt = ChargingPointTypeUtils.HouseholdSocket
+
+    when:
+    def alteredUnit = cpt.copy()
+        .setElectricCurrentType(ElectricCurrentType.DC)
+        .setsRated(Quantities.getQuantity(2.4, StandardUnits.S_RATED))
```
Comment (danielfeismann) [inline src/test/groovy/edu/ie3/datamodel/models/input/system/type/chargingpoint/ChargingPointTypeTest.groovy:22]: This should be 2.3, or?

File: src/test/groovy/edu/ie3/datamodel/models/input/system/type/chargingpoint/ChargingPointTypeTest.groovy [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        .build()
+
+    then:
+    alteredUnit.with {
+      assert id == cpt.id
+      assert sRated == Quantities.getQuantity(2.4, StandardUnits.S_RATED)
```
Comment (danielfeismann) [inline src/test/groovy/edu/ie3/datamodel/models/input/system/type/chargingpoint/ChargingPointTypeTest.groovy:28]: same here

**Review by danielfeismann — CHANGES_REQUESTED** [review]
This looks very good to me. Just the capex question and one minor thing.

File: src/test/groovy/edu/ie3/datamodel/models/input/system/type/chargingpoint/ChargingPointTypeTest.groovy (hunk 1 above)
Comment (sebastian-peter) [reply src/test/groovy/edu/ie3/datamodel/models/input/system/type/chargingpoint/ChargingPointTypeTest.groovy:22]: The household socket has 2.3 kVA originally (as per `ChargingPointTypeUtils`), and to check whether the copy builder works, we change that value to 2.4 kVA and test whether that value has been applied to the new `ChargingPointType`.

https://github.com/ie3-institute/PowerSystemDataModel/blob/667680f35d97170f0c83563c0323bcaf72a07bae/src/main/java/edu/ie3/datamodel/models/input/system/type/chargingpoint/ChargingPointTypeUtils.java#L40-L45

**Comment by sebastian-peter** [comment]
> A more general question. How should we deal with capex? I guess we should scale it for all models as well even this does not represent economy of scale...

I think we should, just to approximate the change. If one looked closely, `opex` could also reduce while scaling up, but for those precise model changes, one should just properly re-model their scenario.

File: src/test/groovy/edu/ie3/datamodel/models/input/system/type/chargingpoint/ChargingPointTypeTest.groovy (hunk 1 above)
Comment (danielfeismann) [reply src/test/groovy/edu/ie3/datamodel/models/input/system/type/chargingpoint/ChargingPointTypeTest.groovy:22]: True. It didn't get the setsRated..

**Review by danielfeismann — APPROVED** [review]
LGTM, thanks for quickly including capex as well.

**Pull request merged**